webExtensionApis/study/src/index.js

```diff
@@ -0,0 +1,395 @@
+/* eslint-env commonjs */
+/* eslint no-logger: off */
+/* global ExtensionAPI */
+
+ChromeUtils.import("resource://gre/modules/ExtensionCommon.jsm");
+ChromeUtils.import("resource://gre/modules/ExtensionUtils.jsm");
+
+const { logger, studyUtils } = require("./studyUtils.js");
+logger.log("in index!");
+
+// eslint-disable-next-line no-undef
+const { EventManager } = ExtensionCommon;
+// eslint-disable-next-line no-undef
+const { EventEmitter, ExtensionError } = ExtensionUtils;
+
+/** Event emitter to handle Events defined in the API
+ *
+ * - onReady
+ * - onEndStudy
+ *
+ */
+class StudyApiEventEmitter extends EventEmitter {
+  emitReady(studyInfo) {
+    this.emit("ready", studyInfo);
+  }
+
+  emitEndStudy(endingResponse) {
+    this.emit("endStudy", endingResponse);
+  }
+}
+
+/** Implements the study/getApi for `browser.study` API */
+this.study = class extends ExtensionAPI {
+  /**
+   * We don't need to override the constructor for other
+   * reasons than to clarify the class member "extension"
+   * being of type Extension
+   *
+   * @param {object} extension Extension
+   */
+  constructor(extension) {
+    super(extension);
+    /**
+     * @type Extension
+     */
+    this.extension = extension;
+    this.studyUtils = studyUtils;
+    this.studyApiEventEmitter = new StudyApiEventEmitter();
+    logger.log("constructed!");
+  }
+
+  /**
+   * Extension Uninstall
+   * APIs that allocate any resources (e.g., adding elements to the browser’s
+   * user interface, setting up internal event listeners, etc.) must free
+   * these resources when the extension for which they are allocated is
+   * shut down.
+   *
+   * https://searchfox.org/mozilla-central/source/toolkit/components/extensions/parent/ext-protocolHandlers.js#46
+   *
+   * @param {string} shutdownReason one of the reasons
+   * @returns {undefined} TODO TODO
+   */
+  async onShutdown(shutdownReason) {
+    logger.log("possible uninstalling", shutdownReason);
+    if (
+      shutdownReason === "ADDON_UNINSTALL" ||
+      shutdownReason === "ADDON_DISABLE"
+    ) {
+      logger.log("definitely uninstalling", shutdownReason);
+      const anEndingAlias = "user-disable";
+      const endingResponse = await this.studyUtils.endStudy(anEndingAlias);
+      await this.studyApiEventEmitter.emitEndStudy(endingResponse);
+    }
+  }
+
+  /**
+   * @param {object} context the addon context
+   * @returns {object} api with study, studyDebug keys
+   */
+  getAPI(context) {
+    const { extension } = this;
+
+    /* eslint no-shadow: off */
+    const { studyUtils, studyApiEventEmitter } = this;
+
+    // once.  Used for pref naming, telemetry
+    studyUtils.setExtensionManifest(extension.manifest);
+    studyUtils.reset();
```

Comment:
What is the purpose of setting this and then resetting?
setExtensionManifest is once again executed in setup()